### ISSUE TITLE
How to extract mAP50  mAP50-95 in each class from model.val() 

### ISSUE THREAD

**Issue by Trirat2543** [ultralytics/ultralytics#7107]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

"How can I extract mAP50 and mAP50-95 for each class from model.val()? Alternatively, if I need to save scores, such as results for plotting, how can I do it and store them in a data frame?
![image](https://github.com/ultralytics/ultralytics/assets/66780756/995e952f-3a4e-4436-9b7d-b887b2d9ad2c)

**Comment by glenn-jocher** [comment]
@Trirat2543 after running `model.val()`, the mAP metrics for each class are automatically computed and displayed in the results. To extract these metrics for your analysis, you can access the output object returned by `model.val()`. This object contains the metrics for each class, and you can programmatically extract the desired values (mAP50, mAP50-95) and store them in a data frame of your choice for further analysis or plotting. Remember to check our documentation at https://docs.ultralytics.com/ for more details on the output structure and how to interact with it. Happy coding! 😊🚀

**Comment by gigumay** [comment]
Hello, I'm a bit confused w.r.t. per-class mAP50 values. As far as I understand the mAP50 value is obtained by averaging the average precision (i.e., area under the precision-recall curve) across classes. So how is the per-class mAP50 value calculated? Is it then simply the average precision for that class? I would find it a bit misleading to call it the "mean" average precision in that case.

**Comment by glenn-jocher** [comment]
@gigumay hello! You're correct that mAP50 is typically the mean of the average precision across all classes. However, when referring to per-class mAP50, it indeed represents the average precision for that specific class at an IoU threshold of 0.50. It's not an average across classes but the performance metric for each individual class. The term "mean" in mAP refers to the average across different IoU thresholds or classes when considering the overall metric. For per-class values, it's simply the average precision at the specified IoU threshold for that class. Hope this clears up the confusion! 😊👍

**Comment by gigumay** [comment]
@glenn-jocher thanks for the explanation!

**Comment by glenn-jocher** [comment]
@gigumay you're welcome! If you have any more questions, feel free to ask. Happy to help! 😊👍

**Comment by paulinavelasquez** [comment]
Hello, I have a question, how can I obtain the instance number and mAP for each class in a data frame when I perform my validation within my training code?:  val=True

**Comment by glenn-jocher** [comment]
Hello @paulinavelasquez,

Thank you for your patience! It sounds like you're on the right track. For YOLOv9, the structure of the results object might differ slightly. To extract the instance count and mAP metrics for each class, you can access the relevant attributes directly from the results object. Here's a more detailed example to help you out:

```python
import pandas as pd
from ultralytics import YOLO

# Load your model
model = YOLO('yolov9.pt')

# Perform validation
results = model.val()

# Extract class-wise metrics
class_metrics = results.results_dict['metrics/per_class']

# Prepare data for DataFrame
data = {
    'class': [],
    'instances': [],
    'mAP50': [],
    'mAP50-95': []
}

for cls, metrics in class_metrics.items():
    data['class'].append(cls)
    data['instances'].append(metrics['instances'])
    data['mAP50'].append(metrics['mAP50'])
    data['mAP50-95'].append(metrics['mAP50-95'])

# Create DataFrame
df = pd.DataFrame(data)
print(df)
```

This code snippet assumes that `results.results_dict['metrics/per_class']` contains the per-class metrics. Adjust the keys as necessary based on the actual structure of your results object.

If you encounter any issues or need further assistance, feel free to ask. We're here to help! 😊

**Comment by paulinavelasquez** [comment]
Thanks for your answer. I tell you that `results` are only returning these metrics for `model = YOLO('yolov9e.pt')`

![image](https://github.com/ultralytics/ultralytics/assets/41736073/b61c7f7c-c5f3-47ae-9291-54625e279760)

**Comment by glenn-jocher** [comment]
Hello @paulinavelasquez,

Thank you for the update and the screenshot! It looks like the `results` object is providing a summary of the overall metrics, but not the detailed per-class metrics directly.

To extract per-class metrics such as the number of instances and mAP for each class, you can dig a bit deeper into the `results` object. Here’s a more detailed approach to help you get those metrics into a DataFrame:

```python
import pandas as pd
from ultralytics import YOLO

# Load your model
model = YOLO('yolov9e.pt')

# Perform validation
results = model.val()

# Extract class-wise metrics
class_metrics = results.results_dict['metrics/per_class']

# Prepare data for DataFrame
data = {
    'class': [],
    'instances': [],
    'mAP50': [],
    'mAP50-95': []
}

# Iterate through the class metrics to populate the data dictionary
for cls, metrics in class_metrics.items():
    data['class'].append(cls)
    data['instances'].append(metrics['instances'])
    data['mAP50'].append(metrics['mAP50'])
    data['mAP50-95'].append(metrics['mAP50-95'])

# Create DataFrame
df = pd.DataFrame(data)
print(df)
```

This code assumes that `results.results_dict['metrics/per_class']` contains the detailed metrics for each class. If the structure is different, you might need to adjust the keys accordingly.

If you still face issues, please share more details about the structure of the `results` object, and we can further refine the solution. We're here to help! 😊

**Comment by paulinavelasquez** [comment]
Hello, I get results from: 
```
results = model.val(imgsz=640,
                    batch=8,
                    conf=0.01,
                    iou=0.7, 
                    save_json=True,
                    save_hybrid=False,
                    split='val')
```
Although I saw that those values ​​are returned in the terminal when executing the `model.train` with `val=true`, then should model itself have those values?

**Comment by glenn-jocher** [comment]
Hello @paulinavelasquez,

Thank you for your question! It’s great to see you diving deep into the validation metrics. You’re correct that the `model.val()` method should provide detailed metrics, including per-class metrics, which can be accessed programmatically.

When you run `model.val()`, the results object should contain all the necessary metrics. However, if you’re not seeing the detailed per-class metrics directly, you might need to explore the structure of the `results` object a bit more.

Here’s a more detailed approach to extract the instance count and mAP metrics for each class:

```python
import pandas as pd
from ultralytics import YOLO

# Load your model
model = YOLO('yolov9e.pt')

# Perform validation
results = model.val(imgsz=640, batch=8, conf=0.01, iou=0.7, save_json=True, save_hybrid=False, split='val')

# Check the structure of the results object
print(results.results_dict.keys())

# Assuming the metrics are stored in 'metrics/per_class'
class_metrics = results.results_dict.get('metrics/per_class', {})

# Prepare data for DataFrame
data = {
    'class': [],
    'instances': [],
    'mAP50': [],
    'mAP50-95': []
}

# Iterate through the class metrics to populate the data dictionary
for cls, metrics in class_metrics.items():
    data['class'].append(cls)
    data['instances'].append(metrics['instances'])
    data['mAP50'].append(metrics['mAP50'])
    data['mAP50-95'].append(metrics['mAP50-95'])

# Create DataFrame
df = pd.DataFrame(data)
print(df)
```

This code snippet assumes that the per-class metrics are stored under `results.results_dict['metrics/per_class']`. If the structure is different, you might need to adjust the keys accordingly.

Additionally, if you’re running `model.train()` with `val=True`, the validation metrics should indeed be part of the model’s attributes. You can access them similarly after training completes.

Feel free to explore the `results` object to find the exact structure and adjust the code as needed. If you have any more questions or need further assistance, don’t hesitate to ask. Happy coding! 😊

**Comment by paulinavelasquez** [comment]
Thank you very much, but rather sorry for the repetitive messages about the problem. I tell you that precisely that metric you mention does not return `'metrics/per_class'`, it currently shows me like this:
![image](https://github.com/ultralytics/ultralytics/assets/41736073/38f97a5e-4f7d-416f-8364-21a451360dba)

Now, I managed to access the metrics by class in the following way `results.box.p[class_index]`, however I still couldn't find the `instance`.

**Comment by glenn-jocher** [comment]
Hello @paulinavelasquez,

No worries at all about the repetitive messages—it's important to get to the bottom of this! 😊

Thank you for sharing the structure of your `results` object. Given that `results.box.p[class_index]` provides access to the per-class metrics, we can work with that to extract the mAP values. However, for the instance count, we might need to dig a bit deeper.

Here’s a refined approach to extract both the mAP metrics and the instance counts:

1. **Extract mAP Metrics:**
   Since you’ve found that `results.box.p[class_index]` gives you the precision, you can similarly access other metrics like recall and mAP.

2. **Extract Instance Counts:**
   If the instance counts are not directly available in the `results` object, you might need to count the instances manually from the validation dataset.

Here’s a code example to illustrate this:

```python
import pandas as pd
from ultralytics import YOLO

# Load your model
model = YOLO('yolov9e.pt')

# Perform validation
results = model.val(imgsz=640, batch=8, conf=0.01, iou=0.7, save_json=True, save_hybrid=False, split='val')

# Assuming class names are available in results.names
class_names = results.names

# Prepare data for DataFrame
data = {
    'class': [],
    'instances': [],
    'mAP50': [],
    'mAP50-95': []
}

# Iterate through each class index
for class_index, class_name in class_names.items():
    data['class'].append(class_name)
    data['mAP50'].append(results.box.map50[class_index])
    data['mAP50-95'].append(results.box.map[class_index])
    
    # Count instances for each class in the validation dataset
    instance_count = sum([1 for label in results.labels if label == class_index])
    data['instances'].append(instance_count)

# Create DataFrame
df = pd.DataFrame(data)
print(df)
```

In this example:
- We iterate through each class index to extract the mAP metrics.
- We count the instances for each class by iterating through the validation labels.

This should give you a comprehensive DataFrame with the class names, instance counts, and mAP metrics.

If you need further assistance or run into any issues, feel free to reach out. We're here to help! 🚀

**Comment by paulinavelasquez** [comment]
Yes, I just created a function to do it manually. Only since it was displayed in the terminal I thought maybe I could get them in variables. Thank you very much for the quick responses. Greetings.

**Comment by glenn-jocher** [comment]
Hello @paulinavelasquez,

Thank you for your patience and for sharing your progress! It's great to hear that you've created a function to manually extract the metrics. While the metrics are displayed in the terminal, accessing them programmatically can indeed be more efficient and flexible.

To help you further, here's a refined approach to extract the mAP metrics and instance counts directly from the `results` object, ensuring you have everything you need in variables:

```python
import pandas as pd
from ultralytics import YOLO

# Load your model
model = YOLO('yolov9e.pt')

# Perform validation
results = model.val(imgsz=640, batch=8, conf=0.01, iou=0.7, save_json=True, save_hybrid=False, split='val')

# Assuming class names are available in results.names
class_names = results.names

# Prepare data for DataFrame
data = {
    'class': [],
    'instances': [],
    'mAP50': [],
    'mAP50-95': []
}

# Iterate through each class index
for class_index, class_name in class_names.items():
    data['class'].append(class_name)
    data['mAP50'].append(results.box.map50[class_index])
    data['mAP50-95'].append(results.box.map[class_index])
    
    # Count instances for each class in the validation dataset
    instance_count = sum([1 for label in results.labels if label == class_index])
    data['instances'].append(instance_count)

# Create DataFrame
df = pd.DataFrame(data)
print(df)
```

This code snippet should help you programmatically extract the metrics and instance counts, making your workflow smoother and more automated.

If you have any further questions or need additional assistance, feel free to ask. We're here to help! 😊

**Comment by paulinavelasquez** [comment]
thanks for the guide.

**Comment by glenn-jocher** [comment]
Hello @paulinavelasquez,

Thank you for your kind words! 😊 We're glad you found the guide helpful.

To ensure we can assist you effectively, could you please provide a minimum reproducible code example? This will help us understand the context and reproduce any issues you might be encountering. You can refer to our guide on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please make sure you are using the latest versions of `torch` and `ultralytics`. You can upgrade your packages using the following commands:

```bash
pip install --upgrade torch
pip install --upgrade ultralytics
```

If you have any specific questions or run into any issues, feel free to share more details, and we'll be happy to assist you further. 

Looking forward to your response!
